### PR TITLE
[cellular][android] Refactor cellular module

### PR DESCRIPTION
**Pull request by M1ST4KE**
# Why

- Modules written in Java are being rewritten to Kotlin
- Module didn't support 5G cellular
- Module was using deprecated android methods

# How

- Generated Kotlin code from java.
- Clean up code.
- Add 5G to enums in Kotlin, Objective-C and Swift
- Add a new method to acquire all information in favour of static fields

# Test Plan

- Checked in the bare-expo app.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).

### CODE REVIEW

File: packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      carrierInfo.putSerializable("mobileCountryCode", null)
+      carrierInfo.putSerializable("mobileNetworkCode", null)
+      carrierInfo.putSerializable("generation", null)
+    }
+    promise.resolve(carrierInfo)
+  }
+
+  @SuppressLint("MissingPermission")
+  private fun getNetworkGeneration(): Int {
+    if (checkPermissions()) {
```
Comment (ide) [inline packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt:104]: The early return reduces nesting and inlines the one line of code:
```suggestion
    if (!mPermissions.hasGrantedPermissions(Manifest.permission.READ_PHONE_STATE)) {
      return CellularGeneration.UNKNOWN.value
    }
```

File: packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      carrierInfo.putSerializable("carrier", telephonyManager.simOperatorName)
+      carrierInfo.putSerializable("mobileCountryCode", telephonyManager.simOperator.take(3))
+      carrierInfo.putSerializable("mobileNetworkCode", StringBuilder(telephonyManager.simOperator).delete(0, 3).toString())
+      try {
+        carrierInfo.putSerializable("generation", getNetworkGeneration())
+      } catch (e: SecurityException) {
+        promise.reject("E_MISSING_PERMISSIONS", e)
+      }  catch (e: Exception) {
+        promise.reject("ERR_CELLULAR_GENERATION_UNKNOWN_NETWORK_TYPE", "Unable to access network type or not connected to a cellular network", e)
+      }
```
Comment (ide) [inline packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt:92]: It seems better to leave this out. That way developers don't need to request permission if they only want to read the carrier and country information.

**Review by ide — CHANGES_REQUESTED** [review]
[One commit should contain one idea](https://github.com/expo/expo/blob/master/guides/Git%20and%20Code%20Reviews.md#one-idea--one-commit). This is a good candidate for four PRs: rewrite in Kotlin (keep feature rewrites separate from technical rewrites), add 5G to the enum, change the constants to methods, and add the permission API for getting the cellular generation.

File: packages/expo-cellular/ios/EXCellular/CellularModule.swift [hunk 3]
```diff
@@ -32,30 +36,69 @@ public class CellularModule: Module {
...
-         CTRadioAccessTechnologyCDMAEVDORev0,
-         CTRadioAccessTechnologyCDMAEVDORevA,
-         CTRadioAccessTechnologyCDMAEVDORevB,
-         CTRadioAccessTechnologyeHRPD:
-      return .cellular3G
-    case CTRadioAccessTechnologyLTE:
-      return .cellular4G
-    default:
-      return .unknown
+    if #available(iOS 14.1, *) {
```
[the remark this answers is not in the document]
Comment (tsapeta) [reply packages/expo-cellular/ios/EXCellular/CellularModule.swift:45]: Unfortunately yes, this enum case is marked as unavailable on older iOS versions and the version we build against doesn't matter 🤷‍♂️ However, I'd recommend moving this `if` into the `default` case, we don't need two switches to handle that.

**Pull request closed without merging**